pkg/templates/externalccm/vsphere.go

```diff
@@ -0,0 +1,317 @@
+/*
+Copyright 2019 The KubeOne Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package externalccm
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"k8c.io/kubeone/pkg/clientutil"
+	"k8c.io/kubeone/pkg/state"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	vSphereSAName           = "cloud-controller-manager"
+	vSphereDeploymentName   = "vsphere-cloud-controller-manager"
+	vSphereConfigSecretName = "cloud-config" //nolint:gosec
+	vSphereImageRegistry    = "gcr.io"
+	vSphereImage            = "/cloud-provider-vsphere/cpi/release/manager:v1.2.1"
+)
+
+func ensurevSphere(s *state.State) error {
+	if s.DynamicClient == nil {
+		return errors.New("kubernetes client not initialized")
+	}
+
+	if s.Cluster.CloudProvider.CloudConfig == "" {
+		return errors.New("cloudConfig not defined")
+	}
+
+	ctx := context.Background()
+
+	sa := vSphereServiceAccount()
+	cr := vSphereClusterRole()
+
+	image := s.Cluster.RegistryConfiguration.ImageRegistry(vSphereImageRegistry) + vSphereImage
+
+	k8sobjects := []runtime.Object{
+		sa,
+		vSphereConfigMap(s.Cluster.CloudProvider.CloudConfig),
+		cr,
+		vSphereClusterRoleBinding(),
+		vSphereRoleBinding(),
+		vSphereDaemonSet(image),
+		vSphereService(),
+	}
+
+	withLabel := clientutil.WithComponentLabel(ccmComponentLabel)
+	for _, obj := range k8sobjects {
+		if err := clientutil.CreateOrUpdate(ctx, s.DynamicClient, obj, withLabel); err != nil {
+			return errors.Wrapf(err, "failed to ensure vSphere CCM %T", obj)
+		}
+	}
+
+	return nil
+}
+
+func vSphereServiceAccount() *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      vSphereSAName,
+			Namespace: metav1.NamespaceSystem,
+		},
+	}
+}
+
+func vSphereConfigMap(cloudConfig string) *corev1.ConfigMap {
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      vSphereConfigSecretName,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Data: map[string]string{
+			"vsphere.conf": cloudConfig,
+		},
+	}
+}
```

Comment:
Why do we keep cloudConfig in a configMap instead of a Secret? It may contain secrets (e.g. if you specify credentials in the plain text), so I think we should use a Secret instead.

Reply:
Good point, normally you would use the  vsphere-ccm-credentials secret by kubeone but it is possible to use plaintext. I changed it to a secret to rule out any possibility.